### ISSUE TITLE
GPS Satellite rework

### ISSUE THREAD

**Issue by Punsho** [OpenRA/OpenRA#17269]
GPS Satellite is a long standing balance issue (not a major one though) and there have been countless discussions about it. Here's my spin on it, so far there have been no complaints and only praise. The exact values likely to be tweaked once it is possible to play test

### GPS Satellite
- Launch timer from 8 to 2 minutes
- Change GPS launch timer to not be displayed globally
   - It would make Allied Tech Centre rush possible to hide, it addresses a common complaint
- Turn GPS into a reusable support power like Spy Plane
   - Once used a “GPS launched” sound would play for the opponent team, for your side it would reveal the whole map with gps dots for 25 seconds. It would recharge in 3 minutes and become ready to be used again. The 3 minute timer would be displayed globally. 
   - The current GPS is flawed in that counterplay options once it is up in the air are very poor and is not great for the Allied player too, it feels like map hacking, it mostly removes the need for scouting and game sense
- Remove Dome as a dependency from GPS
   - It will be awkward to use GPS and for it to not work

### Allied Tech Centre
- Health from 60000 to 80000
   - It would be the same as of Soviet Tech Centre. As GPS is not going to be as powerful there's no need to have it this vulnerable

**Comment by GraionDilach** [comment]
Depends on #16000.

@Punsho AS provides code which is functional for your usecase and more (conditional GPS dots as well, so say GapGens disabling underlying GPS along).

**Comment by netnazgul** [comment]
I'm not sure 3 minute timer should be displayed globally, as it becomes a "minor" support power like Iron Curtain or Chronosphere, both of which don't display their timers for opponents.

Also probably launch timer itself is not required either.

**Comment by matjaeck** [comment]
I like how the current GPS works. In the original, where no fog of war existed, GPS unveiled the whole map. Showing the GPS icons underneath the added fog of war is IMO one of the best improvements, a very well done example of how to modernize the game and still being faithful. The proposed logic would in my perception break with this, without fixing the point of

> counterplay options once it is up in the air are very poor

This point would still apply, just not permanently. There was still no counterplay besides "waiting" until the power is deactivated. I'm not sure if this "3 minute action window" would improve gameplay. It could aswell affect game dynamic negatively. If you don't show the re-charge countdown you have to assume it's up all time. That's even less a counter but source for frustration and uncertainty.

> and is not great for the Allied player too

I ompletely disagree with this - having the GPS "map hack" is super useful. How can a map hack that only lasts for short time be better for the allied player ?

I'm more open towards the idea to hide the countdown if this helped to encourage tech play in 1v1s. Not a fan of further changes to the original tech tree.

**Comment by GraionDilach** [comment]
In any case, the GPS leaks through the Gap Generator.

That's a clear bug to me from gameplay standpoint, caused by inconsistency: it clears the shroud to fog-of-war in RA1 with the activation. If it clears nongenerated shroud then either it should disable GapGens altogether either shouldn't propagate GPS dots through generated shroud or shouldn't remove shroud altogether.

**Comment by GraionDilach** [comment]
Could be a hardcode I disabled during my local copy during making it conditional then, I don't remember now.

**Comment by Krokoschoko** [comment]
>It could aswell affect game dynamic negatively. If you don't show the re-charge countdown you have to assume it's up all time. That's even less a counter but source for frustration and uncertainty.

In the proposal it reads:

>Once used a “GPS launched” sound would play for the opponent team, for your side it would reveal the whole map with gps dots for 25 seconds. 

So you do know when it's being used, even if you don't have an exact timer. It adds no more uncertainty than the possibility of someone having a Chrono shift ready.

**Comment by Punsho** [comment]
A sound notification for when gps disables would also be helpful. Though I'm not sure we have on that could fit.

Now that I think about it. With FoW OFF gps should only reveal the map once by clearing shroud. Further usage would be completely redundant. So for polish we'd have to split the rules for different options. This is why having them is detrimental in the first place.

> counterplay options once it is up in the air are very poor
>> This point would still apply, just not permanently. There was still no counterplay besides "waiting" until the power is deactivated. I'm not sure if this "3 minute action window" would improve gameplay.

It is much easier to kill the tech centre when your opponent doesn't see your every move.

>and is not great for the Allied player too
>>I completely disagree with this - having the GPS "map hack" is super useful. How can a map hack that only lasts for short time be better for the allied player ?

It is useful. It basically removes Fog of War, it removes much of the uncertainty from the game. It is indeed easier to play with perfect information. It might be fun for casual play but is very detrimental to competitive play. It is better for Allied players in a way that it doesn't let you shut off your brain and only act upon what you see at this very moment. You would still need plan and rely on educated guesses. It also help people improve in teamgames as they would be forced to predict enemy movement / actively scout instead of fully relying on gps, it would also not be as big of an issue to not have an Ally in your team. It would also improve lategame of allied mirrors

**Comment by Punsho** [comment]
What we have currently is a mechanic that is by no doubt hugely overpowered. We tried to balance it by making it come up as late in the game as possible and gave a long window of opportunity to stop it from happening. That hasn't worked out well. GPS Satellite only exists because it existed in the originals, not because it adds interesting gameplay. If it weren't for that reason it would have been removed a long time ago. We actually have a decent example in ora ra which deals well with powerful support powers locked behind long timers, Nuke. It is similar but it actually improves gameplay, It helps with breaking stalemates and It discourages clumping up buildings and defences very tightly. It has easy and good counter play, just split your forces and buildings apart, or kill the missile silo. There can be only one silo and if it is killed the timer is reset. Unlike Satellite whichs timer doesn't even reset if launched, if not launched it is possible to have multiple tech centres to keep from resetting. What makes counter play even harder. If nuke hits painfully it not game over yet. The game hasn't been altered permanently. It only just grants a temporary advantage that significantly helps in breaking a stalemate. If you have all your forces split up to counter it. It just means that the nuke improved gameplay by simply existing. It is much  more fun to fight with multiple armies instead of having just one.

**Comment by matjaeck** [comment]
> > It could aswell affect game dynamic negatively. If you don't show the re-charge countdown you have to assume it's up all time. That's even less a counter but source for frustration and uncertainty.
> 
> In the proposal it reads:
> 
> > Once used a “GPS launched” sound would play for the opponent team, for your side it would reveal the whole map with gps dots for 25 seconds.

Thanks for the hint, I misread this. Thinking of bigger team games, I think it makes sense that iron curtain or chronosphere usage is exposed to enemy players by a sound effect rather than an announcement, otherwise it could become annoying. So that might be good for such a feature too.

> It is much easier to kill the tech centre when your opponent doesn't see your every move.

Right, I see what you meant with countering GPS now.

> It basically removes Fog of War, it removes much of the uncertainty from the game. It is indeed easier to play with perfect information. It might be fun for casual play but is very detrimental to competitive play. 

I see the potential in the proposal and think it's possible that competitive play and allied mirrors could profit from the change, provided the games reach tier 3. On the other hand, there are teamgames (and probably singleplayer games) that most of the time go into lategame with (team-)GPS being an important part of it which will definitely be affected.

>  It also help people improve in teamgames as they would be forced to predict enemy movement / actively scout instead of fully relying on gps.

Teamgames are not all the same and this might work well for the regular teamplayers who know what they're doing. But I think a lot of other players do not want to be forced to this and play allies just because of GPS. The people who play the game to relax, often older people and likely with memories for the original, would need to get used to a new game mechanic that didn't make the game any more fun for them. 

Now, to please the oldtimers and purists can of course not be the only scale to measure new changes in. But since the change would be quite disruptive, it's a factor to consider, just like the potential improvement for high skilled players and viewers.

This leads me to the conclusion that the time for a new "permanent GPS" lobby checkbox might has come.

> Unlike Satellite whichs timer doesn't even reset if launched, if not launched it is possible to have multiple tech centres to keep from resetting.

Related to #16647 for reset options and #10359 for restarting.

**Comment by Punsho** [comment]
Only allies would be able to infiltrate and reset if #16647 was merged. And allies already have a few counters, GG and MGG while soviets have... planes? And about #10359. Allies can still just make multiple tech centres to deny snipes (also fakes exist). And FYI most of the time killing the tech centre is expensive. 

>But I think a lot of other players do not want to be forced to this and play allies just because of GPS

To put it into perspective. In this release in team games most people play soviet. Why? Because mammoths are overpowered and they are easy to use. Why do some people still play as allies? Because GPS is overpowered and is easy to use. Of course there are many exceptions but these are the trends.

**Comment by GraionDilach** [comment]
>Only allies would be able to infiltrate and reset if #16647 was merged. And allies already have a few counters, GG and MGG while soviets have... planes?

Assuming that gets merged, I guess there could be a discussion to extend Thief's infiltration to support that case but still not supporting Barracks/fakes/whatelse.

**Comment by matjaeck** [comment]
@Punsho FWIW, you don't need to convince me. When you said "there have been no complaints and only praise" I felt challenged to add another perspective. Maintainers seem to stay neutral however and I'm not interested to start another fight with the discord horde behind you.

**Comment by airetaM** [comment]
Another option is to reset the timer if a player looses all tech centers.
the timer should be shorter then of course and maybe not be displayed in the upper left.

**Comment by Punsho** [comment]
> I'm not sure 3 minute timer should be displayed globally, as it becomes a "minor" support power like Iron Curtain or Chronosphere, both of which don't display their timers for opponents.

About that. Opponents can see when iron curtain is in effect, same goes for chonoshift. But with gps you can only estimate if it's still online. Or maybe you didn't hear the notification and want to check if it's seeing you. The timer is needed for that specific reason. After use it would reset to 3:25

**Comment by pchote** [comment]
IMO have two different timer states would be awkward.

For the sake of discussion I'll throw in a competing idea:
* Keep the same initial countdown and launch behaviour
* Make it a targeted power that reveals dots in a limited area, and is always active (no secondary timer)
* Player can select a new target from the support power palette, but it takes time for the field of view to move around the map (revealing the path as it goes)
* Keep requirement on an active radar
* Remove all public timers and notification sounds

**Comment by praetp** [comment]
> IMO have two different timer states would be awkward.
> 
> For the sake of discussion I'll throw in a competing idea:
> 
>     * Keep the same initial countdown and launch behaviour
> 
>     * Make it a targeted power that reveals dots in a limited area, and is always active (no secondary timer)
> 
>     * Player can select a new target from the support power palette, but it takes time for the field of view to move around the map (revealing the path as it goes)
> 
>     * Keep requirement on an active radar
> 
>     * Remove all public timers and notification sounds

I like it, it's closer to the behavior of an actual spy satellite.

**Comment by matjaeck** [comment]
I also think it is an interesting idea. What would be the initial position of the the spy sat? IMO it should still automatically go online and not only after the power is used.

**Comment by Punsho** [comment]
Spy plane-like satellite has an inherent problem, on the receiving end you can't know where it is. As I've mentioned before in various places, if you don't know that you're being scouted, you lose. It is as simple as that. All you can do is hope that Eye of Sauron is distracted

As for double timers. Two problems clash here and I currently see double timer as the only good solution. GPS is very powerful thus it needs to be very clear of when it is online. That means a global timer is needed. Since GPS Satellite is launched from Allied Tech Centre showing global timer immediately makes it impossible to hide Allied tech rush and surprise your opponents

**Comment by pchote** [comment]
Surely knowing the enemy may be able to see a small area but not knowing where it is is a significant improvement over knowing that the enemy can see everywhere all the time?

**Comment by GraionDilach** [comment]
Just place a beacon in the middle of the revealed area IMO if that's a big deal.

**Comment by Punsho** [comment]
> Surely knowing the enemy may be able to see a small area but not knowing where it is is a significant improvement over knowing that the enemy can see everywhere all the time?

It is very similar. With the current gps your opponent may still not see a few moves because he didn't move his vision port there. But you still have no way of knowing that he didn't scout and the risk is usually too great to commit an attack (when the game is at a stalemate)

> Just place a beacon in the middle of the revealed area IMO if that's a big deal.

Is looking for and avoiding a bunch of dash line circles really the gameplay we want to have?

**Comment by matjaeck** [comment]
Just keep everything as it then and call it a won't fix. If people don't like Red Alert, they should change the game they play but not Red Alert should change for them.

We can't please everybody, and need to stay faithful to the original. Even pchotes idea introduces an alien concept that is unknown to the original games, and OP insists to alienate the game even more.

**Comment by matjaeck** [comment]
And on a side note: if people fail to scout that their enemy rushed a TC, or fail to react to or prevent it, then there is an issue that the player needs to solve but not the game developers. So I actually don't see any reason to expose tech paths to the enemy other than helping players that fail to scout. It says quite a lot about the state of competitive OpenRA when people insist the game needs to help them more... If GPS is such a gamebreaker, why do only so few "competitive" games reach T3? Why is the mobile gap generator not considered super valuable then?

The reason the competitive level stagnates is that players don't need to come up with solutions for challenging game mechanics but instead argue to change the game. It's ridiculous.

**Comment by Punsho** [comment]
> Just keep everything as it then and call it a won't fix. If people don't like Red Alert, they should change the game they play but not Red Alert should change for them.
> 
> We can't please everybody, and need to stay faithful to the original. Even pchotes idea introduces an alien concept that is unknown to the original games, and OP insists to alienate the game even more.


> And on a side note: if people fail to scout that their enemy rushed a TC, or fail to react to or prevent it, then there is an issue that the player needs to solve but not the game developers. So I actually don't see any reason to expose tech paths to the enemy other than helping players that fail to scout. It says quite a lot about the state of competitive OpenRA when people insist the game needs to help them more... If GPS is such a gamebreaker, why do only so few "competitive" games reach T3? Why is the mobile gap generator not considered super valuable then?
> 
> IMO the reason why the competitive level in my eyes stagnates is that players don't need to come up with solutions for challenging game mechanics but instead argue to change the game.

This is just a long argument to play the original. OpenRA supposed to be better. Look at the competitive games in CNCnet. Certain units are just unviable, so is the reason for it that the players have gotten too good or that the game is imbalanced?

**Comment by matjaeck** [comment]
You misunderstood - I meant OpenRA's Red Alert, not the original game. These mechanics are part of OpenRA's Red Alert, they are the result of what I hope is a faithful and careful modernization of the original game. OpenRA's Red Alert is supposed to be better than the original *where it can be* - but it is still Red Alert and not a game we can do with what we want.

> Certain units are just unviable

Then we need to fix them as much as we can. If the units and the mechanics are after all what we were able to do (and I can get behind a lot of things) still imbalanced and badly designed, then sorry, understand the game is from 1996 and accept it is in some areas badly designed. What you call imbalanced and badly designed is what others call authentic and faithful.

**Comment by Punsho** [comment]
I didn't misunderstand. I was trying to give you more perspective on what you were proposing. I advise to reread what you wrote again

> they are the result of what I hope is a faithful and careful modernization of the original game

This is the process. This is how we produce these changes. This is how we modernise. It isn't helpful to just shut it down. With statements like this "Just keep everything as it then and call it a won't fix. If people don't like Red Alert, they should change the game they play but not Red Alert should change for them."

>  If GPS is such a gamebreaker, why do only so few "competitive" games reach T3

I didn't mention this before. Medium skill games don't reach it often (unless over 4 people are fighting). Between the top players most games reach radar and at the minimum one quarter (much more often for some players) of those reach T3. I wouldn't call that a small number.

**Comment by Naigel56** [comment]
Hello everybody, sorry for chiming in. I can't understand why people want to penalize one of the most interesting allies features. In the current meta, it doesn't seem to be needed.

Btw, if we want to make the TC more interesting, I remember another idea popped up in the past, we can consider it in this discussion. Basically, it consists in limiting the range of the tech center. This will make the TC a bit more tactical and it changes its use since players may want to build more than one. It affects more the multiplayer/big maps meta where the advantage of the TC is arguably bigger (all the team benefits at once from one).
To counterbalance, a reduction in the countdown seems to be a good solution. It may not be easy to find the perfect equilibrium between the countdown range and the visibility radius. I would consider 3 options:

- High countdown/range: 4 minutes + 25-30 cells
- Medium: 2 minutes + 20-25 cells
- low: no countdown + 16 cells

**Comment by SoScared** [comment]
> IMO have two different timer states would be awkward.
> 
> For the sake of discussion I'll throw in a competing idea:
> 
> * Keep the same initial countdown and launch behavior
> * Make it a targeted power that reveals dots in a limited area, and is always active (no secondary timer)
> * Player can select a new target from the support power palette, but it takes time for the field of view to move around the map (revealing the path as it goes)
> * Keep requirement on an active radar
> * Remove all public timers and notification sounds

I had to comment. This looks like a great idea, if I still modded I'd immediately steal it 😃 

I'd imagine a big global visible range circle for where the GSP is active, making it a tactical tool both player and opponent would have to deal with. Use in multiplayer could be a lot of fun as well.

**Comment by wippie-openra** [comment]
I love the suggestions in the opening statement. I would like to add to make Hide GPS timer to be faction specific (France) would be better.

**Comment by al-dyachenko** [comment]
better return original red alert 1 aftermath balance completely and dont make any s***

**Comment by Pockets-byte** [comment]
I'm a big fan of the idea of a permanent "search area" as if it's an actual unit 
it makes it less blatantly a "explore the whole map button" in shroud only games, it makes fog of war games much less tilted towards allied, and generally fixes many of the issues I have with gps both playing as and playing against
 this is a bit of a stretch- but perhaps a dedicated expensive satellite interceptor unit that is unlocked with high tech and can shoot down a gps sat when in it's field of view for a certain amount of time when playing against allies might be a good counterplay for poking sauron in the eye.  sorta like something like the the anti-nuke in supcomm/TAlikes
